src/go/rpk/pkg/serde/proto.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package serde
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/protocompile"
+	"github.com/twmb/franz-go/pkg/sr"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// EncodeProtoRecord encodes the provided record using the specified schema.
+// It utilizes the message type identified by protoFQN. If the schema includes
+// references, it retrieves them using the supplied client. This function
+// returns the record encoded in the protobuf wire format.
+func EncodeProtoRecord(ctx context.Context, cl *sr.Client, schema *sr.Schema, protoFQN string, schemaID int, record []byte) ([]byte, error) {
+	if protoFQN == "" {
+		return nil, fmt.Errorf("please provide the protobuf message name")
+	}
+	// Compile the Schema Registry proto. //
+	const inMemFileName = "tmp.proto"
+	accessorMap := make(map[string]string)
+	// This is the original schema.
+	accessorMap[inMemFileName] = schema.Schema
+
+	// And we add the rest of schemas if we have references.
+	if len(schema.References) > 0 {
+		err := mapReferences(ctx, cl, schema, accessorMap)
+		if err != nil {
+			return nil, fmt.Errorf("unable to map references: %v", err)
+		}
+	}
+
+	compiler := protocompile.Compiler{
+		Resolver: &protocompile.SourceResolver{
+			Accessor: protocompile.SourceAccessorFromMap(accessorMap),
+		},
+		SourceInfoMode: protocompile.SourceInfoStandard,
+	}
+	compiled, err := compiler.Compile(ctx, inMemFileName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compile the given schema: %v", err)
+	}
+
+	// Create proto message with runtime information. //
+	d := compiled.FindFileByPath(inMemFileName).FindDescriptorByName(protoreflect.FullName(protoFQN))
+	if d == nil {
+		return nil, fmt.Errorf("unable to find message with name %q", protoFQN)
+	}
+	msgDescriptor, ok := d.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("unable to process message with name %q", protoFQN)
+	}
+	message := dynamicpb.NewMessage(msgDescriptor)
+
+	// Unmarshal the record into the proto message. //
+	o := protojson.UnmarshalOptions{
+		Resolver: compiled.AsResolver(),
+	}
```

Comment:
I think everything from the top of this function to this line should only run once -- currently this doing a lot of logic for every record when we only need to do it once at the start.

What do you think about introducing an encoder type into this package,
```go
type Encoder struct {
    fn func([]byte) error
}
```
and then _all_ of this logic can either run before `fn` or inside `fn`, captured. When combined with my comment below, this can turn into:

```go
// schema already has had it's references resolved, so we do not need ctx nor srCl
func NewEncoder(schema *sr.Schema, schemaID int, protoFQN string) (*Encoder, error) {
    switch schema.Type {
    case avro: // simpler case, but same thing!
    case proto:
        encFn, err := newProtoEncoder(schema, schemaID, protoFQN)
        return ...
    }
}

func newProtoEncoder(...) ... {
  // lines 29 through 71
  return func(record []byte) error {
    // lines 72 through 86
  }, nil
}
```

Reply:
I took a peek at the consume branch as well -- this comment applies for building a decoder as well. Maybe the type can be named Serder, the func NewSerde, and then this function builds the encoder and decoder at the same time (since the setup is _mostly_ the same).

Reply:
PTAL at https://github.com/redpanda-data/redpanda/compare/291bc40a6f826619022eafc0dcd87a2ee35359c2..5729c8ba8d0b43f6c74eaf7c50182fed8c00f450 and resolve if you agree with the approach :+1: